psi4/driver/driver_cbs.py

```diff
@@ -703,6 +893,149 @@ def return_energy_components():
 
 VARH = return_energy_components()
 
+def _get_default_xtpl(nbasis, xtpl_type):
+    """ A helper function to determine default extrapolation type.
+    
+    Parameters
+    ----------
+    nbasis : int
+        Number of basis sets
+    xtpl_type : str
+        Extrapolation type ("scf" or "corl").
+    
+    Returns
+    -------
+    function
+        Extrapolation function to be used.
+    """
+    
+    if nbasis == 1:
+        return xtpl_highest_1
+    elif xtpl_type ==  "scf":
+        if nbasis == 2:
+            return scf_xtpl_helgaker_2
+        elif nbasis == 3:
+            return scf_xtpl_helgaker_3
+        else:
+            raise ValidationError("Wrong number of basis sets supplied to scf_xtpl: %d" % nbasis)
+    elif xtpl_type ==  "corl":
+        if nbasis == 2:
+            return corl_xtpl_helgaker_2
+        else:
+            raise ValidationError("Wrong number of basis sets supplied to corl_xtpl: %d" % nbasis)
+            
+
+def _process_cbs_kwargs(kwargs):
+    """ A helper function which either translates supplied kwargs into the
+    ``cbs_metadata`` format, or validates a provided cbs_metadata array.
+    
+    Parameters
+    ----------
+    kwargs : dict
+        kwargs containing the CBS function specification.
+    
+    Returns
+    -------
+    array
+        Array of dictionaries, with each item consisting of an extrapolation
+        stage. All validation takes place here.
+    """    
+    
+    metadata = []
+    if "cbs_metadata" in kwargs:
+        cbs_metadata = kwargs.get("cbs_metadata")
+        for item in cbs_metadata:
+            metadata.append({})
+            if "wfn" in item:
+                if "wfn" not in item:
+                    raise ValidationError("Wavefunction type has to be supplied for every stage.")
+                else:
+                    metadata[-1]["wfn"] = item.get("wfn")
+                if "basis" not in item:
+                    raise ValidationError("Basis set has to be supplied for every stage.")
+                else:
+                    metadata[-1]["basis"] = _expand_bracketed_basis(item.get("basis").lower())
+                metadata[-1]["treatment"] = item.get("treatment", "scf" if len(metadata) == 1 else "corl")
+                metadata[-1]["scheme"] = item.get("scheme", _get_default_xtpl(len(metadata[-1]["basis"][1]),
+                                                                                 metadata[-1]["treatment"])) 
+                if len(metadata) > 1:
+                    metadata[-1]["wfn_lo"] = item.get("wfn_lo", metadata[-2].get("wfn"))
+                    metadata[-1]["basis_lo"] = _expand_bracketed_basis(item.get("basis_lo", 
+                                                                       item.get("basis")).lower())
+                    if len(metadata[-1]["basis"][0]) != len(metadata[-1]["basis_lo"][0]):
```

Comment:
I wonder if it'd be a bit cleaner to have a `trial_stage = {}` at the start, fill that instead of `metadata[-1]`, then `metadata.append(trial_stage)` at the end if all successful. Open for debate.

Reply:
Fine by me.